### PR TITLE
gnrc_ipv6_nib: fix acquire race on gnrc_ipv6_nib_get_next_hop_l2addr()

### PR DESCRIPTION
**Pull request by miri64**
### Contribution description
When two threads use `gnrc_ipv6_nib_get_next_hop_l2addr()` to determine a next hop (e.g. when there is both an IPv6 sender and a 6LoWPAN fragment forwarder), a race condition may happen, where one thread acquires the NIB and the other acquires the network interface resulting in a deadlock. By releasing the NIB (if acquired) before trying to acquire the network interface and re-acquiring the NIB after the network interface is acquired, this is fixed.

An example for how the race condition might come to pass is that first, the IPv6 thread, acquires both the `netif` and the NIB for a look-up of an off-link (global) address:

https://github.com/RIOT-OS/RIOT/blob/6e4434a76049356d5c24be606084cd38abc6d432/sys/net/gnrc/network_layer/ipv6/nib/nib.c#L198-L199

and then continues to https://github.com/RIOT-OS/RIOT/blob/6e4434a76049356d5c24be606084cd38abc6d432/sys/net/gnrc/network_layer/ipv6/nib/nib.c#L261-L269

However, when the 6LoWPAN thread (with a fragment forwarding module like `gnrc_sixlowpan_frag_minfwd` or `gnrc_sixlowpan_frag_sfr` via the VRB [here]()) during that also makes a lookup but with a specific `netif != NULL` (same as the new `netif` in IPv6) in L198, it will get stuck on L199, since the IPv6 thread still hold the lock on the NIB, while then waiting on the `netif` in 268. I did not really observe this with the link-local variant

https://github.com/RIOT-OS/RIOT/blob/6e4434a76049356d5c24be606084cd38abc6d432/sys/net/gnrc/network_layer/ipv6/nib/nib.c#L210-L215

but in theory it can also happen there.

### Testing procedure
Both `tests/gnrc_ipv6_nibu` and `tests/gnrc_ipv6_nib_6ln` should still pass and `gnrc_networking` should still work. I found this by running experiments very similar to https://github.com/5G-I3/IEEE-LCN-2019/, but with a higher send interval (250-750ms), for SFR with congestion control. Without this fix I ran into the race condition quite often with low congestion windows. With this fix, I was not able to reproduce the race condition yet.

### Issues/PRs references
None.

### CODE REVIEW

**Comment by miri64** [comment]
I started the [`release-test`](https://github.com/RIOT-OS/RIOT/actions/runs/814134342) with this PR. If there is a problem with this PR, it should show up there.

File: sys/net/gnrc/network_layer/ipv6/nib/nib.c [hunk 1]
```diff
@@ -186,6 +186,18 @@ static bool _on_link(const ipv6_addr_t *dst, unsigned *iface)
     return ipv6_addr_is_link_local(dst);
 }
 
+static gnrc_netif_t *_acquire_new_iface(unsigned iface)
```
Comment (benpicco) [inline sys/net/gnrc/network_layer/ipv6/nib/nib.c:189]: Why not

```suggestion
static gnrc_netif_t *_acquire_new_iface(gnrc_netif_t *netif)
```

Looks like that when you are calling this, you already have a pointer to `netif`.

Comment (miri64) [reply sys/net/gnrc/network_layer/ipv6/nib/nib.c:189]: Nope, as you can see, I am getting the `netif` pointer from the numerical `iface` below first. Sure, I can do that also before calling `_acquire_new_iface()`, but my thought was, if I do it in any case, why not do it in this function?

Comment (benpicco) [reply sys/net/gnrc/network_layer/ipv6/nib/nib.c:189]: In line 225 you already have the `netif` pointer

Comment (miri64) [reply sys/net/gnrc/network_layer/ipv6/nib/nib.c:189]: To clarify: This function is used in the case when the provided `netif` is not the one the next hop is associated with. So, the `netif` I "already have", is a different one potentially, than the one that is returned by this function. That's why I need to release and then acquire `netif` again.

Comment (miri64) [reply sys/net/gnrc/network_layer/ipv6/nib/nib.c:189]: > In line 225 you already have the `netif` pointer

Look at what the comment says: "release pre-assumed netif". This `netif` pointer is a different one from the one we fetch in the line below. The `if` condition in l272 should make clear what I mean.

**Review by benpicco — APPROVED** [review]
Looks good and makes sense. I see no potential for regressions here.

**Comment by miri64** [comment]
> Looks good and makes sense. I see no potential for regressions here.

As this is supposed to be backported, lets at least wait for https://github.com/RIOT-OS/RIOT/runs/2511278567.

**Comment by miri64** [comment]
There is one failing task in https://github.com/RIOT-OS/RIOT/actions/runs/814134342, but that's a LoRaWAN test and thus unrelated.

**Comment by miri64** [comment]
Backport provided in #16452

**Pull request merged**